### ISSUE TITLE
[BUG] "Second" backlog after rescheudling even though I've already rescheduled all cards before

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki#438]
**Describe the bug**
I remember someone on Reddit mentioning this issue as well, and now I have encountered it myself. Recently, probably one or two weeks ago, I rescheduled all my cards using FSRS and got a huge backlog of around 2000 cards. Today I rescheduled them again out of curiosity, and got a huge backlog (~1000 cards) again. Just to add, I'm still not completely done with the previous backlog either.
This is quite urgent and important, as this could be a sign of a major problem with the scheduler/helper add-on.

**To Reproduce**
1. Go to Tools > FSRS4Anki and click "Reschedule all cards", see a large backlog
2. Go through the backlog
3. Do your reviews normally
4. Repeat step 1
5. See a large backlog again

**Expected behavior**
The first backlog makes sense, since I've been using Anki for almost 1000 days and rescheduled to FSRS. But getting a backlog when rescheduling from FSRS to FSRS after 1-2 weeks doesn't make any sense.

**Environment**
 - Anki version: 2.1.64
 - OS: Windows 10
 - Scheduler version: 4.3.0

**Additional context**
https://www.reddit.com/r/Anki/comments/15ggz5d/rescheduling_with_fsrs_continues_to_result_in/ another user with the same issue.
[Here](https://drive.google.com/file/d/1-0RWzX1DVoBJAhRiU53eOCA3IZopnY2D/view?usp=sharing) is my collection, as well as my FSRS settings. Note that my learning and relearning steps are all set to 1 day max, I meticulously checked the preset of every single deck.

**Comment by user1823** [comment]
Even a learning (or relearning) step of 1 day would cause some issues. Read this: https://github.com/open-spaced-repetition/fsrs4anki/issues/351#issuecomment-1685358240

However, the above-mentioned issue is not likely to cause a backlog. It will most likely decrease your due count after rescheduling.

Another possible cause: https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/173#issuecomment-1658204186

**Comment by L-M-Sherlock** [comment]
Weird. Could you compare the graphs of `Future Due` in the stats before and after rescheduling?

**Comment by Expertium** [comment]
![1](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/01cca0c0-d2a8-4ffe-a21d-85353ebcfb8a)

**Comment by Expertium** [comment]
![Before-after](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/527cfe54-ed98-4e05-82ea-30dce294406c)

**Comment by L-M-Sherlock** [comment]
@Expertium, when did you set `"maximumInterval": 365`?

<img width="800" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/5b191daa-6a16-4803-a075-88a1f32f4ce1">

I find that this card's interval is 5.19 years. And you set "maximumInterval": 365` for that deck:

```
  {
    // Deck 12
    "deckName": "Main::Science",
    "w": [1.75, 0.34, 1.05, 25.12, 4.3769, 0.999, 0.9592, 0.0, 1.5168, 0.1063, 0.9033, 2.1476, 0.1321, 0.3643, 0.9035, 0.2433, 3.2814],
    "requestRetention": 0.8,
    "maximumInterval": 365,
  },
```

**Comment by Expertium** [comment]
Ah, yeah, I find that deck to be pretty difficult overall, so I decided to set it low.
However, other decks exhibit the same behavior. Go to Main > English > Hard English 
```
"deckName": "Main::English::Hard English",
    "w": [0.77, 0.19, 0.95, 61.14, 4.6653, 2.0528, 1.2786, 0.0001, 1.8413, 0.1641, 1.2478, 2.2611, 0.1288, 0.3966, 1.7925, 0.4036, 3.8385],
    "requestRetention": 0.8,
    "maximumInterval": 1825,
```
Here the max. interval is 5 years, but I still get a second backlog.

**Comment by L-M-Sherlock** [comment]
<img width="1064" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/7345a1c7-784c-473d-86f5-8c4acb30e545">

The card's stability doesn't change significantly, but its interval is changed a lot after rescheduling (1.69y -> 9.7m).

You requestRetention is 0.8. So the interval = 9 * 136 * (1/0.8 - 1) = 306 (~ 9.7m). Did you increase the requestRetention after your first rescheduling?

**Comment by Expertium** [comment]
> Did you increase the requestRetention after your first rescheduling?

No.

**Comment by Expertium** [comment]
> Do you enable load balance in the last rescheduling?

I always have it enabled.

**Comment by L-M-Sherlock** [comment]
> This is quite urgent and important, as this could be a sign of a major problem with the scheduler/helper add-on.

I find that most of these cards are not reviewed in weeks. So it's not the problem of the scheduler.

**Comment by Expertium** [comment]
Maybe it has something to do with Postpone, but I don't think I have postponed that many cards recently.

**Comment by Expertium** [comment]
Also, I don't think that the difference in interval length with and without postponing would be that big.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/767f48de-16dd-4e84-9910-9c8f2942ff19)

**Comment by Expertium** [comment]
So @L-M-Sherlock, any ideas what is going here and how to solve it?

**Comment by L-M-Sherlock** [comment]
I haven't figured out which module went wrong. I need more experiments or feedback about that phenomenon.

**Comment by jake0216** [comment]
I have had this experience as well, been using FSRS since March and have optimized each month. Every time I get a backlog of 200-600 cards. I have noticed this will happen if I do not reschedule with the FSRS addon daily as well. 

I have saved my weights from each optimization and have backups of my deck at each optimization time point as well, let me know if any of that would be helpful

**Comment by L-M-Sherlock** [comment]
The main problem is the interval is wrong but the stability is correct.

**Comment by L-M-Sherlock** [comment]
To locate the bug, I need to store the scheduler info in the custom data, such as postpone/advance/disperse.

**Comment by L-M-Sherlock** [comment]
I add the source of schedule in this PR.
- https://github.com/open-spaced-repetition/fsrs4anki-helper/pull/204

<img width="452" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/f499ecc7-2454-479a-93c1-1a379ffcbe1a">
<img width="450" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/b448c59f-a943-473e-b00d-e79b2544ffa6">
<img width="468" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/99658781-1030-47d6-9ab1-016c764a414c">
<img width="486" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/7cafbd47-9a3b-4aa9-b506-b065a8f3e04c">

It will help us to locate the bug.

**Comment by L-M-Sherlock** [comment]
The cause may be `disperse`. I find that I forget to apply max interval in `disperse`. I have fixed it in https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/857cba9a70adf62473c07ebc7b8f17f7d416facb.

**Comment by Expertium** [comment]
My bad, I didn't test it properly. Ideally, I should've tried rescheduling before the update and after the update to see the difference. But I only tried rescheudling after the update.
It did give me some extra review cards today, but since I haven't tried it before the update, I don't know whether the number of cards would be higher or lower without the update.

**Comment by Expertium** [comment]
Ok, I was still able to find something interesting. The cards that became due after rescheduling have all been rescheduled normally, without the sibling dispersion.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/badd99a7-5632-45d4-a696-9c5152751044)
This raises the question - why did they become due only after rescheduling? As I have previously mentioned, I have already rescheduled all my cards.

**Comment by user1823** [comment]
I still think that [this](https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/173#issuecomment-1658204186) can be a cause.

**Comment by L-M-Sherlock** [comment]
> I still think that [this](https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/173#issuecomment-1658204186) can be a cause.

If those cards due after rescheduling are caused by disperse siblings, their original card info would include v=disperse. We can check it.

**Comment by Expertium** [comment]
On the left you can see the info before rescheduling, on the right you can see the info after rescheduling.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/353b0a89-7704-45a4-8511-3fb4764e99dd)

**Comment by L-M-Sherlock** [comment]
Could you include the intervals?

By the way, v=helper means it is rescheduled by old version of the helper.

**Comment by L-M-Sherlock** [comment]
For convenience of test, I plan to add a new column for `v`.

**Comment by user1823** [comment]
> For convenience of test, I plan to add a new column for `v`.

In the browser?

I don't think that it would be very useful. If you want to search for cards with v=disperse or something like that, you can use `prop:cdn`.

**Comment by L-M-Sherlock** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/f42e7551-9804-4efd-9f71-d4fecd7c794a)

😢

**Comment by Expertium** [comment]
> Could you include their intervals?

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/5e0c628d-87d9-498f-a7f4-62c182bf6de7)

**Comment by L-M-Sherlock** [comment]
1.76 year = 642 day
11.47 month = 344 day

I guess the first one doesn't respect your maximum interval, which has been solved in https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/857cba9a70adf62473c07ebc7b8f17f7d416facb.

**Comment by Expertium** [comment]
My max interval for that deck is 1825 days

**Comment by L-M-Sherlock** [comment]
(1 + 642/9/145.47)^-1 = 0.67

(1 + 344/9/145.47)^-1 = 0.79

That is weird. Your request retention is 0.8, right? Could you check this card's info? Maybe the problem is caused by disperse.

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/9e9ed050-f509-407f-aa1b-40ad3ec65d3a)

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/6ed226dc-e3b4-4a72-8fdd-eba9f447c356)

**Comment by L-M-Sherlock** [comment]
OK, I will check the code of disperse siblings tomorrow.

**Comment by L-M-Sherlock** [comment]
The bug is too hard to find out. I develop a patch to show the info of disperse siblings:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/e58d57b0-85f4-452e-ba20-c1d3c46d5b95)

To enable it, you need to set it to true (false by default):

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/a30c1385-7309-48cc-bbfd-9e04c2d6df91)

**Comment by L-M-Sherlock** [comment]
If you find that the due data change too much, you can share the affected card's info here.

**Comment by L-M-Sherlock** [comment]
What about displaying the target retention in the browser?

**Comment by Expertium** [comment]
That may help

**Comment by Expertium** [comment]
I have already found outliers.
I have a card with Target R=4.59%, and a lot of cards with Target R<80%, I never set it that low for any deck. I have also found cards with Target R>95%, and again, I never set it that high for any decks. Some cards have N/A despite having been reviewed in the past. If you want me to, I can put all outliers into a special deck and send it to you.

**Comment by L-M-Sherlock** [comment]
I also found these cases. The card with Target R = 4.59% is has a manual reset in its last review. And cards with Target R>95% are reaching maximum interval. Do you have found any other cases?

**Comment by L-M-Sherlock** [comment]
I think FSRS should respect the interval set by manual reset and maximum interval, because manual reset and maximum interval are intended.

**Comment by Expertium** [comment]
> Do you have found any other cases?

No, I found only these 3 cases:
1) Cards with target R<80%
2) Cards with target R>95%
3) Cards with at least one review and Target R=N/A

**Comment by L-M-Sherlock** [comment]
Which one is unexpected? You can share screenshots here and I will check them tomorrow.

**Comment by Expertium** [comment]
> Which one is unexpected?

All of them? I mean, none of those 3 things should happen.

**Comment by L-M-Sherlock** [comment]
> I think FSRS should respect the interval set by manual reset and maximum interval, because manual reset and maximum interval are intended.

Please consider my point above.

**Comment by Expertium** [comment]
By the way, how is target R calculated?

**Comment by L-M-Sherlock** [comment]
> Why not just find which deck a card belongs to, then find the corresponding `requestRetention` value from the scheduler code?

Because I need to calculate the true target retrievability, which could be affected by load balacne and dispersing. If I use the requestRetention value, we couldn't find out the bug proposed in current issue.

**Comment by Expertium** [comment]
Another user who seems to have a similar problem: https://reddit.com/r/Anki/s/2MJQC8XtHa

By the way, I still haven't found anything that could help us solve this.

**Comment by user1823** [comment]
> Another user who seems to have a similar problem: https://reddit.com/r/Anki/s/2MJQC8XtHa

This is not the same problem. Here, the parameters are different (due to re-optimization). But, in your case, the parameters are the same.

The case in that reddit comment is similar to https://github.com/open-spaced-repetition/fsrs4anki/issues/187. I don't think that we can do anything to solve it.

However, I think that this problem would become less severe if the algorithm becomes even more accurate (e.g. in v5 or v6).

My understanding is that this problem occurs because FSRS is unable to "accurately" predict the R for some cases (e.g. high D, post-lapse, etc.). Thus, when the newer data is fed to the optimizer, it tries to readjust the weights so as to decrease the loss but is still unable to predict future R accurately (because the formulas are not accurate), repeating the story after the next optimization.

That being said, I am very grateful to @L-M-Sherlock for developing this wonderful algorithm and also for making it open-source. Even if it is not "very" accurate, it is still much better than SM-2 and probably SM-17 also. I hope that it becomes even more accurate in the future.

**Comment by Expertium** [comment]
I'm not 100% sure, but I think this problem doesn't exist in beta6.
I just tried the following with one of my decks:
1) Change desired retention from 0.80 to 0.81 with "Reschedule on change" enabled
2) The number of due cards increased
3) I changed desired R back to 0.8
4) The number o due cards decreased to the **exact** same number as before these steps
So unless my testing has a flaw, it seems like this is a problem with the standalone version of FSRS.

**Comment by L-M-Sherlock** [comment]
Does the problem exist in the helper add-on for Anki 2.1.66 or below?

**Comment by Expertium** [comment]
> Does the problem still exist in the helper add-on for Anki 2.1.66 or below?

I don't think I can test it. I am now using per-preset parameters rather than per-deck parameters (and I have combined a few presets and deleted a few others), so they are different.
Btw, the deck I tested was rescheduled using built-in FSRS several days ago. So it was "built-in FSRS" -> "built-in FSRS" kind of rescheduling. It's worth mentioning because in order to test it properly, I need to be consistent.

**Comment by L-M-Sherlock** [comment]
Due to the lack of reports from other users, I think it's OK to close this issue?

**Comment by Expertium** [comment]
I guess so.

**Comment by Expertium** [comment]
@L-M-Sherlock since this issue is closed, it should be unpinned
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/3ab38041-6ade-4e08-9ca9-8a510c764388)
